Archive/Imo/Imo2024Q5.lean

```diff
@@ -0,0 +1,1065 @@
+/-
+Copyright (c) 2024 Joseph Myers. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joseph Myers
+-/
+import Mathlib.Data.Fin.VecNotation
+import Mathlib.Data.List.ChainOfFn
+import Mathlib.Data.Nat.Dist
+import Mathlib.Order.Fin.Basic
+import Mathlib.Tactic.IntervalCases
+
+/-!
+# IMO 2024 Q5
+
+Turbo the snail plays a game on a board with $2024$ rows and $2023$ columns. There are hidden
+monsters in $2022$ of the cells. Initially, Turbo does not know where any of the monsters are,
+but he knows that there is exactly one monster in each row except the first row and the last
+row, and that each column contains at most one monster.
+
+Turbo makes a series of attempts to go from the first row to the last row. On each attempt,
+he chooses to start on any cell in the first row, then repeatedly moves to an adjacent cell
+sharing a common side. (He is allowed to return to a previously visited cell.) If he reaches a
+cell with a monster, his attempt ends and he is transported back to the first row to start a
+new attempt. The monsters do not move, and Turbo remembers whether or not each cell he has
+visited contains a monster. If he reaches any cell in the last row, his attempt ends and the
+game is over.
+
+Determine the minimum value of $n$ for which Turbo has a strategy that guarantees reaching
+the last row on the $n$th attempt or earlier, regardless of the locations of the monsters.
+
+We follow the solution from the
+[official solutions](https://www.imo2024.uk/s/IMO2024-solutions-updated.pdf). To show that $n$
+is at least $3$, it is possible that the first cell Turbo encounters in the second row on his
+first attempt contains a monster, and also possible that the first cell Turbo encounters in the
+third row on his second attempt contains a monster. To show that $3 attempts suffice, the first
+attempt can be used to locate the monster in the second row; if this is not at either side of
+the board, two more attempts suffice to pass behind that monster and from there go along its
+column to the last two, while if it is at one side of the board, the second attempt follows a
+zigzag path such that if it encounters a monster the third attempt can avoid all monsters.
+-/
+
+
+namespace Imo2024Q5
+
+/-! ### Definitions for setting up the problem -/
+
+-- There are N monsters, N+1 columns and N+2 rows.
+variable {N : ℕ}
+
+/-- The board for the game. -/
+abbrev Board (N : ℕ) : Type := Fin (N + 2) × Fin (N + 1)
```

Comment:
Would it make more sense to name this `Cell` (or possibly `Coords`)? The name `Board` gives me the wrong idea that this type will contain information about what occupies *all* of the cells.

Reply:
I've renamed this to `Cell`.